File: SecondYearCS/OOP/CivitasRB/lib/dado.rb
  require 'singleton'
module CivitasRB
    class Dado
    include Singleton #Singleton
    
    def initialize(salida_carcel=5) #Constructor con parámetros.
       @debug = false
       @ultimo_resultado = 0
       @@SalidaCarcel = salida_carcel
    end
    
    #Generamos un número aleatorio entre 1-6 con rand
    #ultimoResultado guarda ese resultado
    def tirar
      if !@debug
        @ultimo_resultado = rand(1..6)
      else
        @ultimo_resultado = 1
      end
      @ultimo_resultado
    end
    
    #Verificamos si el ultimoResultado es mayor que lo necesario para salr 
    #De la carcel.
    def salgoDeLaCarcel
      self.tirar
      salir = false
      if @ultimo_resultado >= @@SalidaCarcel
        salir = true
      end
      return salir
    end
    
    #Generamos un número aleatorio entre 0-(n-1) para saber quien empieza.
    def quienEmpieza(n)
      return rand(n)
    end
    
    #Empezamos a depurar nuestro dado.
    #Además escribimos el evento en nuestro diario.
    def setDebug(d)
      @debug = d
      Diario.instance.ocurre_evento("Debug is activated in Dado")
    end
    
    def ultimoResultado
      return @ultimoResultado
    end
    
  end
end 

File: SecondYearCS/OOP/OldStuff/Dado.java
package OldStuff;
import java.util.Random;

class Dado {

    static final private Dado instance = new Dado(); //implementar singleton
    
    private final Random random;
    private int ultimoResultado;
    private boolean debug;
    private static int SalidaCarcel = 5; // valor 5 como static const al principio
    
    static public Dado getInstance(){
        return instance;
    }
    //Constructor sin parámetros
    private Dado(){
        this.random = new Random();
        this.ultimoResultado = 0;
        this.debug = false;
    }
    /**
     * @brief Devuelve un número aleatorio generado entre 1-6 ( Dado )
     * @return @b this.ultimoResultado ( parámetro de la clase )
     */
    int tirar(){
        this.ultimoResultado = 1;
        if(!this.debug){
            this.ultimoResultado = this.random.nextInt(6)+1;
        }
        return this.ultimoResultado;
    }
    /**
     * @brief Verificamos si el @b this.ultimoResultado >= @b this.SalidaCarcel
     * @return verdadero/falso
     */
    boolean salgoDeLaCarcel(){
        boolean result = false;
        if(this.ultimoResultado >= Dado.SalidaCarcel)
            result = true;
        return result;
    }
    /**
     * @brief Decide que jugar empieza utilizando this.random
     * @param n número de jugadores
     * @return el jugador que debe empezar entre 0 y n-1
     */
    int quienEmpieza(int n){
        return this.random.nextInt(n);
    }
    /**
     * @brief Cambiamos el valor de @b this.debug
     * @param d nuevo valor de @b this.debug
     */
    public void setDebug(boolean d){
        this.debug = d;
        Diario.getInstance().ocurreEvento("Debug");
    }
    /**
     * @return devuelve el @b this.ultimoResultado
     */
    int getUltimoResultado(){
        return this.ultimoResultado;
    }
    
}
